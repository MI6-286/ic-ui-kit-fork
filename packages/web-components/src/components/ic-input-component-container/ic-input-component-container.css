ic-input-component-container {
  /**
   * @prop --border-color: Border colour of the input component container
   * @prop --border-color-hover: Border colour of the input component container on hover
   * @prop --border-color-pressed: Border colour of the input component container when pressed
   * @prop --border-color-disabled: Border colour of the input component container when disabled
   * @prop --border-color-error: Border colour of the error variant for input component container
   * @prop --border-color-error-hover: Border colour of the error hover variant for input component container
   * @prop --border-color-error-pressed: Border colour of the error pressed variant for input component container
   * @prop --border-color-warning: Border colour of the warning variant for input component container
   * @prop --border-color-warning-hover: Border colour of the warning hover variant for input component container
   * @prop --border-color-warning-pressed: Border colour of the warning pressed variant for input component container
   * @prop --border-color-success: Border colour of the success variant for input component container
   * @prop --border-color-success-hover: Border colour of the success hover variant for input component container
   * @prop --border-color-success-pressed: Border colour of the success pressed variant for input component container
   */

  display: flex;
  border: var(--ic-border-width) solid
    var(--border-color, var(--ic-architectural-400));
  border-radius: var(--ic-border-radius);
  transition: var(--ic-easing-transition-slow);
  height: 2.5rem;
  min-height: 2.5rem;
  width: var(--input-width, 20rem);
  padding: var(--ic-space-1px);
  background-color: var(--input-bg-color, var(--ic-color-background-primary));
  box-sizing: border-box;
  fill: var(--ic-architectural-400);
  outline: none;
}

ic-input-component-container.ic-input-component-container-full-width {
  width: 100%;
}

ic-input-component-container.ic-input-component-container-disabled,
ic-input-component-container.ic-input-component-container-disabled:hover {
  border: var(--ic-border-width) dashed
    var(--border-color-disabled, var(--ic-color-border-neutral-disabled));
}

ic-input-component-container.ic-input-component-container-readonly,
ic-input-component-container.ic-input-component-container-small.ic-input-component-container-readonly,
ic-input-component-container.ic-input-component-container-large.ic-input-component-container-readonly {
  height: auto;
}

ic-input-component-container.ic-input-component-container-readonly,
ic-input-component-container.ic-input-component-container-readonly:hover {
  border: none;
  padding: 0;
}

ic-input-component-container.ic-input-component-container-error {
  border: var(--ic-space-xxxs) solid
    var(--border-color-error, var(--ic-color-border-error));
  padding: 0;
}

ic-input-component-container.ic-input-component-container-error:hover {
  border: var(--ic-space-xxxs) solid
    var(--border-color-error-hover, var(--ic-color-border-error-hover));
}

ic-input-component-container.ic-input-component-container-error:focus-within {
  border: var(--ic-space-xxxs) solid
    var(--border-color-error-pressed, var(--ic-color-border-error-pressed));
}

ic-input-component-container.ic-input-component-container-warning {
  border: var(--ic-space-xxxs) solid
    var(--border-color-warning, var(--ic-color-border-warning));
  padding: 0;
}

ic-input-component-container.ic-input-component-container-warning:hover {
  border: var(--ic-space-xxxs) solid
    var(--border-color-warning-hover, var(--ic-color-border-warning-hover));
}

ic-input-component-container.ic-input-component-container-warning:focus-within {
  border: var(--ic-space-xxxs) solid
    var(--border-color-warning-pressed, var(--ic-color-border-warning-pressed));
}

ic-input-component-container.ic-input-component-container-success {
  border: var(--ic-space-xxxs) solid
    var(--border-color-success, var(--ic-color-border-success));
  padding: 0;
}

ic-input-component-container.ic-input-component-container-success:hover {
  border: var(--ic-space-xxxs) solid
    var(--border-color-success-hover, var(--ic-color-border-success-hover));
}

ic-input-component-container.ic-input-component-container-success:focus-within {
  border: var(--ic-space-xxxs) solid
    var(--border-color-success-pressed, var(--ic-color-border-success-pressed));
}

ic-input-component-container.ic-input-component-container-small {
  height: var(--ic-space-xl);
  min-height: var(--ic-space-xl);
}

ic-input-component-container.ic-input-component-container-large {
  height: var(--ic-space-xxl);
  min-height: var(--ic-space-xxl);
}

ic-input-component-container.ic-input-component-container-multiline {
  height: auto;
}

ic-input-component-container .icon-container {
  margin-top: var(--ic-space-xxs);
  margin-left: 0.438rem;
  display: flex;
  align-items: center;
}

ic-input-component-container.ic-input-component-container-multiline
  .icon-container,
ic-input-component-container.ic-input-component-container-multiline.ic-input-component-container-small
  .icon-container {
  margin-top: 0.375rem;
  display: block;
}

ic-input-component-container.ic-input-component-container-readonly
  .icon-container {
  margin-left: -0.313rem;
}

ic-input-component-container.ic-input-component-container-disabled
  ::placeholder {
  display: none;
}

ic-input-component-container .inline-success {
  margin: var(--ic-space-xs) 0.375rem;
  display: flex;
  align-items: center;
}

ic-input-component-container .inline-success > svg {
  fill: var(--ic-input-component-container-success-icon-inline-color);
  height: 1.25rem;
  width: 1.25rem;
}

ic-input-component-container:hover {
  border: var(--ic-input-component-container-hover-border);
  border-color: var(--border-color-hover, var(--ic-color-border-neutral-hover));
  color: var(--ic-action-dark-hover);
}

ic-input-component-container:focus {
  border: var(--ic-border-pressed);
  border-color: var(
    --border-color-pressed,
    var(--ic-color-border-neutral-pressed)
  );
}

.focus-indicator {
  display: flex;
  width: 100%;
  margin: -0.125rem;
  padding: 0.125rem;
  border-radius: var(--ic-border-radius);
  transition: var(--ic-easing-transition-fast);
}

.focus-indicator:focus-within,
.focus-indicator-enabled {
  box-shadow: var(--ic-border-focus);
}

.focus-indicator.ic-input-component-container-dark:focus-within {
  box-shadow: var(--ic-border-focus);
}

@media (forced-colors: active) {
  /* Styles here only apply to Forced Colors Mode */
  ic-input-component-container,
  .focus-indicator {
    transition: none;
  }

  ic-input-component-container:focus-within {
    border: var(--ic-border-width) solid Highlight;
    outline: 0.125rem solid Highlight;
  }

  ic-input-component-container.ic-input-component-container-disabled,
  ic-input-component-container.ic-input-component-container-disabled:hover {
    border: var(--ic-border-width) dashed GrayText;
  }
}
