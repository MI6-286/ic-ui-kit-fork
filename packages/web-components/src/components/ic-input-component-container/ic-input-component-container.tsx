import { Component, Element, Host, Prop, Watch, h } from "@stencil/core";

import {
  IcInformationStatus,
  IcInformationStatusOrEmpty,
  IcSizes,
} from "../../utils/types";
import successIcon from "../../assets/success-icon.svg";
import {
  removeDisabledFalse,
  renderDynamicChildSlots,
  slotHasContent,
} from "../../utils/helpers";

/**
 * @slot left-icon - Content will be placed to the left of the input.
 */
@Component({
  tag: "ic-input-component-container",
  styleUrl: "ic-input-component-container.css",
})
export class InputComponentContainer {
  private hostMutationObserver: MutationObserver;
  @Element() el: HTMLIcInputComponentContainerElement;

  /**
   *  If `true`, the disabled state will be set.
   */
  @Prop() disabled: boolean = false;
  @Watch("disabled")
  watchDisabledHandler(): void {
    removeDisabledFalse(this.disabled, this.el);
  }

  /**
   *  If `true`, the input component container will fill the width of the container it is in.
   */
  @Prop() fullWidth: boolean = false;

  /**
   *  If `true`, the input component container will allow for multiple lines.
   */
  @Prop() multiLine: boolean = false;

  /**
   *  If `true`, the readonly state will be set.
   */
  @Prop() readonly: boolean = false;

  /**
   * The size of the input component container component.
   */
  @Prop() size?: IcSizes = "medium";

  /**
   *  If `true`, the validation will display inline.
   */
  @Prop() validationInline: boolean = false;

  /**
   * The validation status of the input component container - e.g. 'error' | 'warning' | 'success'.
   */
  @Prop() validationStatus: IcInformationStatusOrEmpty = "";

  componentWillLoad(): void {
    removeDisabledFalse(this.disabled, this.el);
  }

  componentDidLoad(): void {
    this.hostMutationObserver = new MutationObserver((mutationList) =>
      renderDynamicChildSlots(mutationList, "left-icon", this)
    );
    this.hostMutationObserver.observe(this.el, { childList: true });
  }

  render() {
    const {
      size,
      validationStatus,
      disabled,
      readonly,
      multiLine,
      fullWidth,
      validationInline,
    } = this;

    return (
      <Host
        class={{
          [`ic-input-component-container-${size}`]: true,
          [`ic-input-component-container-${validationStatus}`]:
            validationStatus !== "" && !disabled && !readonly,
          "ic-input-component-container-disabled": disabled,
          "ic-input-component-container-readonly": readonly,
          "ic-input-component-container-multiline": multiLine,
          "ic-input-component-container-full-width": fullWidth,
        }}
        aria-disabled={disabled ? "true" : null}
      >
        <div class="focus-indicator">
          {slotHasContent(this.el, "left-icon") && (
            <div class="icon-container">
              <slot name="left-icon" />
            </div>
          )}
          <slot></slot>

          {validationInline &&
            validationStatus === IcInformationStatus.Success && (
              <span class="inline-success" innerHTML={successIcon} />
            )}
        </div>
      </Host>
    );
  }
}
