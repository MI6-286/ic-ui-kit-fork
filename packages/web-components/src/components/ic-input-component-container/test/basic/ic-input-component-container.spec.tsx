import {
  mockHasDynamicChildSlots,
  mockMutationObserverImplementation,
  MockMutationRecord,
  mockRenderDynamicChildSlots,
} from "../../../../testspec.setup";
import { InputComponentContainer } from "../../ic-input-component-container";
import { newSpecPage } from "@stencil/core/testing";

describe("ic-input-component-container", () => {
  afterAll(() => {
    jest.restoreAllMocks();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container>content</ic-input-component-container>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-input-component-container class="ic-input-component-container-medium">
      <div class="focus-indicator">
        content
      </div>
    </ic-input-component-container>
    `);
  });

  it("should render with validation status", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container validation-status="success">content</ic-input-component-container>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-input-component-container class="ic-input-component-container-medium ic-input-component-container-success" validation-status="success">
      <div class="focus-indicator">
        content
      </div>
    </ic-input-component-container> 
    `);
  });

  it("should render with validation inline", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container validation-status="success" validation-inline=true>content</ic-input-component-container>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-input-component-container class="ic-input-component-container-medium ic-input-component-container-success" validation-inline="true" validation-status="success">
      <div class="focus-indicator">
        content
        <span class="inline-success">
          svg
        </span>
      </div>
    </ic-input-component-container>
    `);
  });

  it("should ignore validation inline if not success", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container validation-status="error" validation-inline=true>content</ic-input-component-container>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-input-component-container class="ic-input-component-container-error ic-input-component-container-medium" validation-inline="true" validation-status="error">
      <div class="focus-indicator">
        content
      </div>
    </ic-input-component-container>
    `);
  });

  it("should render with size small, multiline & fullwidth props set", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container size="small" multi-line=true full-width=true>content</ic-input-component-container>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-input-component-container class="ic-input-component-container-small ic-input-component-container-full-width ic-input-component-container-multiline" full-width="true" multi-line="true" size="small">
      <div class="focus-indicator">
        content
      </div>
    </ic-input-component-container>
    `);
  });

  it("should render with size large", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container size=large>content</ic-input-component-container>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-input-component-container class="ic-input-component-container-large" size="large">
      <div class="focus-indicator">
        content
      </div>
    </ic-input-component-container>
    `);
  });

  it("should render disabled", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container validation-status="success" disabled=true>content</ic-input-component-container>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-input-component-container aria-disabled="true" class="ic-input-component-container-disabled ic-input-component-container-medium" disabled="true" validation-status="success">
      <div class="focus-indicator">
        content
      </div>
    </ic-input-component-container>
    `);

    page.rootInstance.disabled = false;

    await page.waitForChanges();
    expect(page.root).toEqualHtml(`
      <ic-input-component-container class="ic-input-component-container-medium ic-input-component-container-success" validation-status="success">
        <div class="focus-indicator">
          content
        </div>
      </ic-input-component-container>
      `);
  });

  it("should render readonly", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container validation-status="success" readonly=true>content</ic-input-component-container>`,
    });

    expect(page.root).toEqualHtml(`
      <ic-input-component-container class="ic-input-component-container-medium ic-input-component-container-readonly" readonly="true" validation-status="success">
        <div class="focus-indicator">
          content
        </div>
      </ic-input-component-container>
    `);
  });

  it("should render with slotted content", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container validation-status="success" disabled=true><span slot="left-icon">slotted content</span>content</ic-input-component-container>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-input-component-container aria-disabled="true" class="ic-input-component-container-disabled ic-input-component-container-medium" disabled="true" validation-status="success">
      <div class="focus-indicator">
        <div class="icon-container">
          <span slot="left-icon">
            slotted content
          </span>
        </div>
        content
      </div>
    </ic-input-component-container>
    `);
  });

  it("should test rendering icon after initial render", async () => {
    const page = await newSpecPage({
      components: [InputComponentContainer],
      html: `<ic-input-component-container>content</ic-input-component-container>`,
    });

    const component = page.rootInstance;
    const host = page.root;

    const observerInstance =
      mockMutationObserverImplementation.mock.results[0].value;

    const icon = document.createElement("svg");
    icon.setAttribute("slot", "left-icon");

    observerInstance.observe(host, { childList: true });

    host.appendChild(icon);

    const mockMutationRecord: MockMutationRecord[] = [
      {
        addedNodes: [icon],
        removedNodes: [],
        target: host,
      },
    ];

    observerInstance.trigger(mockMutationRecord);

    await page.waitForChanges();

    expect(mockRenderDynamicChildSlots).toHaveBeenCalledTimes(1);
    expect(mockRenderDynamicChildSlots).toHaveBeenCalledWith(
      mockMutationRecord,
      "left-icon",
      component
    );

    expect(mockHasDynamicChildSlots).toHaveBeenCalledTimes(1);

    expect(page.root).toMatchSnapshot();
  });
});
