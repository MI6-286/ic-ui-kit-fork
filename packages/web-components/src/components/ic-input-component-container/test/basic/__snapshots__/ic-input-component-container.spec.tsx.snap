// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-input-component-container should test rendering icon after initial render 1`] = `
<ic-input-component-container class="ic-input-component-container-medium">
  <!---->
  <div class="focus-indicator">
    <div class="icon-container">
      <svg slot="left-icon"></svg>
    </div>
    content
  </div>
</ic-input-component-container>
`;
